### ISSUE TITLE
Error in Function backward when forward output is in-place of forward input and input is not leaf

### ISSUE THREAD

**Issue by garibarba** [pytorch/pytorch#5871]
- OS: Win10
- PyTorch version: '0.3.1.post2'
- How you installed PyTorch (conda, pip, source):  `conda install -c peterjc123 pytorch` 
- Python version: Python 3.6.4 :: Anaconda custom (64-bit)

I'm using in-place addition on slices to save memory. To do that I'm defining my own `Function`.
Here is a minimal example reproducing my problem, and the workaround I've found.

```
import torch
from torch.autograd import Variable, Function

class InPlaceAdd(Function):
    @staticmethod
    def forward(ctx, a, b):
        a[1:2] += b
        return a
    
    @staticmethod
    def backward(ctx, grad_output):
        return grad_output, grad_output[1:2].resize(1)

add_in_place = InPlaceAdd.apply
```

This fails to compute the gradient on `b`:
```
a = Variable(torch.ones(2), requires_grad=True)
b = Variable(torch.ones(1), requires_grad=True)

c = add_in_place(a + 1, b)

c.sum().backward()

print(a.grad, b.grad)
```
* output:
```
Variable containing:
 1
 1
[torch.FloatTensor of size 2]
 None
```

while this works just fine:
```
a = Variable(torch.ones(2), requires_grad=True)
b = Variable(torch.ones(1), requires_grad=True)

c = add_in_place(a, b)

c.sum().backward()

print(a.grad, b.grad)
```
* output:
```
Variable containing:
 1
 1
[torch.FloatTensor of size 2]
 Variable containing:
 1
[torch.FloatTensor of size 1]
```

---------------------------------
The workaround I've found is returning a `view` to the output `Tensor` in `forward` instead of the input `tensor` itself. With this change, the problem is gone.
`return a` => `return a.view_as(a)`

**Comment by apaszke** [comment]
There are a few problems with your implementation. Most notably, you fail to use `ctx.mark_dirty` to signal that you've modified `a`. Secondly, you should use `.view(1)` instead of `.resize(1)`, because the later results in the output having **unspecified** contents. Finally, I tried reproducing your issues on `master`, but it seems to work just fine, so I'm closing the issue as resolved.

**Comment by garibarba** [comment]
Thanks for your answer. `ctx.mark_dirty` is missing from the documentation, but anyway my use of in-place assignment is probably too particular. Thanks for the tip on `resize`, edited accordingly. I will try this again on the next release.
